src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -1289,51 +1288,42 @@ internal static bool IsGenericTaskType(this TypeSymbol type, CSharpCompilation c
             {
                 return true;
             }
-            NamedTypeSymbol builderType;
-            MethodSymbol createBuilderMethod;
-            return namedType.IsCustomTaskType(out builderType, out createBuilderMethod);
+            object builderArgument;
+            return namedType.IsCustomTaskType(out builderArgument);
         }
 
         /// <summary>
-        /// Returns true if the type is generic or non-generic task-like type. If so, the async
-        /// method builder type is returned along with the method to construct that type.
+        /// Returns true if the type is generic or non-generic custom task-like type due to the
+        /// [AsyncBuilder(typeof(B))] attribute. It returns the "B".
         /// </summary>
-        internal static bool IsCustomTaskType(this NamedTypeSymbol type, out NamedTypeSymbol builderType, out MethodSymbol createBuilderMethod)
+        /// <remarks>
+        /// For the Task types themselves, this method might return true or false depending on mscorlib.
+        /// The definition of "custom task-like type" is one that has an [AsyncBuilder(typeof(B))] attribute,
+        /// no more, no less. Validation of builder type B is left for elsewhere. This method returns B
+        /// without validation of any kind.
+        /// </remarks>
+        internal static bool IsCustomTaskType(this NamedTypeSymbol type, out object builderArgument)
         {
             Debug.Assert((object)type != null);
             Debug.Assert(type.SpecialType != SpecialType.System_Void);
 
             var arity = type.Arity;
             if (arity < 2)
             {
-                // Find the public static CreateAsyncMethodBuilder method.
-                var members = type.GetMembers(WellKnownMemberNames.CreateAsyncMethodBuilder);
-                foreach (var member in members)
+                // Find the AsyncBuilder attribute.
+                foreach (var attr in type.GetAttributes())
                 {
-                    if (member.Kind != SymbolKind.Method)
-                    {
-                        continue;
-                    }
-                    var method = (MethodSymbol)member;
-                    if ((method.DeclaredAccessibility == Accessibility.Public) &&
-                        method.IsStatic &&
-                        (method.ParameterCount == 0) &&
-                        !method.IsGenericMethod)
+                    if (attr.IsTargetAttribute(type, AttributeDescription.AsyncBuilderAttribute)
+                        && attr.CommonConstructorArguments.Length == 1
+                        && attr.CommonConstructorArguments[0].Kind == TypedConstantKind.Type)
```

Comment:
Added unit tests in https://github.com/dotnet/roslyn/pull/13405/commits/33e65fbff780e88abaf2fb969e9afafe323e54b1